File: src/chapter02/genericlist_test.c
#include <assert.h>

#define _TESTING
#include "genericlist.c"

int executed = 0;

void spy_print(Item *item)
{
	executed++;
}

int main()
{
	List *scores;
	float first, second, third;

	first = 1.11;
	scores = new_list((void*)&first);
	assert(scores != NULL);
	assert(size(scores) == 1);
	assert(*(float*)get(scores, 0) == first);

	second = 2.22;
	append(scores, (void*)&second);
	assert(size(scores) == 2);
	assert(*(float*)get(scores, 0) == first);
	assert(*(float*)get(scores, 1) == second);

	third = 3.33;
	append(scores, (void*)&third);
	assert(size(scores) == 3);
	assert(*(float*)get(scores, 0) == first);
	assert(*(float*)get(scores, 1) == second);
	assert(*(float*)get(scores, 2) == third);

	assert(executed == 0);
	each(scores, spy_print);
	assert(executed == 3);

	puts("ok");

	return 0;
}

File: src/chapter02/GenericListTest.java
class StringSpyPrinter implements Printer {
    private int executions = 0;
    private StringBuilder output = new StringBuilder();

    public void print(Object value) {
        output.append(value.toString());
        executions++;
    }

    public boolean contains(Object value) {
        return output.toString().contains(value.toString());
    }

    public int getExecutions() {
        return executions;
    }
}

class GenericListTest {
    public static void main(String args[]) {
        List list = new List("first");
        assert list != null;
        assert list.size() == 1;
        assert list.get(0) != null;
        assert list.get(0).getValue().equals("first");

        list.append("second");
        assert list.size() == 2;
        assert list.get(0).getValue().equals("first");
        assert list.get(1).getValue().equals("second");

        list.append("third");
        assert list.size() == 3;
        assert list.get(0).getValue().equals("first");
        assert list.get(1).getValue().equals("second");
        assert list.get(2).getValue().equals("third");

        StringSpyPrinter spy = new StringSpyPrinter();
        list.each(spy);

        assert spy.getExecutions() == 3;
        assert spy.contains("first");
        assert spy.contains("second");
        assert spy.contains("third");

        System.out.println("ok");
    }
}
